common/src/java/org/apache/hadoop/hive/conf/HiveConf.java

```diff
@@ -4509,7 +4509,7 @@ private static void populateLlapDaemonVarsSet(Set<String> llapDaemonVarsSetLocal
         "Minimum allocation possible from LLAP buddy allocator. Allocations below that are\n" +
         "padded to minimum allocation. For ORC, should generally be the same as the expected\n" +
         "compression buffer size, or next lowest power of 2. Must be a power of 2."),
-    LLAP_ALLOCATOR_MAX_ALLOC("hive.llap.io.allocator.alloc.max", "16Mb", new SizeValidator(),
+    LLAP_ALLOCATOR_MAX_ALLOC("hive.llap.io.allocator.alloc.max", "4Mb", new SizeValidator(),
```

Comment:
The issue here is that LLAP_ALLOCATOR_MAX_ALLOC is also used as the ORC Writer buffer size (thus the change).

Initial buffer size check was introduced in [ORC-238](https://github.com/apache/orc/pull/171/files) even though it was only applied when buffer size was enforced from table properties. Later, on ORC-1.6 this was enforced for the [Writer buffer size in general](https://github.com/apache/orc/blob/0128f817b0ab28fa2d0660737234ac966f0f5c50/java/core/src/java/org/apache/orc/impl/WriterImpl.java#L171).

The max bufferSize argument can be up to 2^(3*8 - 1) -- meaning less than 8Mb and since we enforce the size to be power of 2 the next available is 4Mb.

The main question here is if there is a reason for the upper limit to be < 8 Mb (cc @prasanthj that might know more here) -- or if we should decouple the two configuration (LLAP alloc and ORC Writer buffer size).

I believe the best thing to do for now is open a new Ticket to track this (as this will either require more work on LLAP, or a new release on ORC) -- and I do not expect this to cause any major issues until then. @mustafaiman what do you think?

Reply:
I still do not understand why we need to change LLAP Allocator's maximum allocation size. Does LLAP allocator serve only ORC writers? I think it is used for other buffer needs too.

Hive depends on ORC. So I dont understand how ORC uses LLAP_ALLOCATOR_MAX_ALLOC for anything. We pass orc writers the appropriate configs. If ORC writers need smaller buffer, we can configure that for those writers via WriterOptions. There is no need to change llap allocator's settings for that.

Reply:
LLAP_ALLOCATOR_MAX_ALLOC is used both for the LowLevelCacheImpl (buddyAllocator) and bufferSize on [WriterOptions](https://github.com/apache/hive/blob/da1aa077716a65c2a02d850828b16cdeece1f574/llap-server/src/java/org/apache/hadoop/hive/llap/io/encoded/SerDeEncodedDataReader.java#L1553) 

Please check how this propagated from [SerDeEncodedDataReader](https://github.com/apache/hive/blob/da1aa077716a65c2a02d850828b16cdeece1f574/llap-server/src/java/org/apache/hadoop/hive/llap/io/encoded/SerDeEncodedDataReader.java#L248) 

Llap is tightly coupled to ORC, thus it could make sense to use the same buffer size for serialized Buffers, and the ORC writer as we would not need to split/merge them  -- however I have nothing against splitting the conf or checking is the 8Mb limit is a hard one.
All I am trying to say here is that this is orthogonal the ORC version bump.

Reply:
I see ORC strictly enforces this now. I would set the appropriate setting at Hive-ORC boundary and leave the LLAP_ALLOCATOR_MAX_ALLOC as it is (Math.min(llap.allocator.max, what ORC enforces). If you think we should set LLAP_ALLOCATOR_MAX_ALLOC to be the same as what ORC enforces, that can be done in a seperate ticket. Like you said this is orthogonal to ORC version bump, therefore should be discussed in its own ticket.

Reply:
Sure, addressed as part of 2eca3b9de1f2332beabc3bde9ac0f89d62ec1527  -- also opened HIVE-24721 to investigate this further